From 68869fd857cc36dfb78d709de4669105889430c3 Mon Sep 17 00:00:00 2001
From: yao zhang1 <yao.zhang1@amlogic.com>
Date: Thu, 9 Apr 2026 15:26:34 +0800
Subject: [PATCH] fs: Set i_count=0 before evicting the inode. [1/1]

PD#SWPL-219092

Problem:
Warning in fscrypt_destroy_keyrin when shutdown.

Solution:
Set i_count=0 before evicting the inode.

Verify:
t6d

Change-Id: Ie4045cc132e8f68a0a597fd6929fa6ed4b540bd0
Signed-off-by: yao zhang1 <yao.zhang1@amlogic.com>
---
 fs/super.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/fs/super.c b/fs/super.c
index 6a8a1b67a35a..b5cb9f445d0f 100644
--- a/fs/super.c
+++ b/fs/super.c
@@ -40,6 +40,10 @@
 #include <uapi/linux/mount.h>
 #include "internal.h"
 
+#ifdef CONFIG_AMLOGIC_PIN_LOCKED_FILE
+/* the macro must large than enum mapping_flags in include/linux/pagemap.h */
+#define AS_LOCK_MAPPING	10
+#endif
 static int thaw_super_locked(struct super_block *sb, enum freeze_holder who);
 
 static LIST_HEAD(super_blocks);
@@ -618,6 +622,9 @@ EXPORT_SYMBOL(retire_super);
 void generic_shutdown_super(struct super_block *sb)
 {
 	const struct super_operations *sop = sb->s_op;
+#ifdef CONFIG_AMLOGIC_PIN_LOCKED_FILE
+	struct inode *inode;
+#endif
 
 	if (sb->s_root) {
 		shrink_dcache_for_umount(sb);
@@ -626,6 +633,18 @@ void generic_shutdown_super(struct super_block *sb)
 
 		cgroup_writeback_umount(sb);
 
+#ifdef CONFIG_AMLOGIC_PIN_LOCKED_FILE
+		spin_lock(&sb->s_inode_list_lock);
+		list_for_each_entry(inode, &sb->s_inodes, i_sb_list) {
+			if (atomic_read(&inode->i_count) <= 0)
+				continue;
+			/* The file is to be removed, dec ref by mlock */
+			if (test_and_clear_bit(AS_LOCK_MAPPING, &inode->i_mapping->flags)) {
+				atomic_set(&inode->i_count, 0);
+			}
+		}
+		spin_unlock(&sb->s_inode_list_lock);
+#endif
 		/* Evict all inodes with zero refcount. */
 		evict_inodes(sb);
 
-- 
2.46.0

